File: src/net/dinkla/raytracer/objects/mesh/SmoothMeshTriangle.java
package net.dinkla.raytracer.objects.mesh;

import net.dinkla.raytracer.hits.Hit;
import net.dinkla.raytracer.math.*;
import net.dinkla.raytracer.objects.mesh.Mesh;
import net.dinkla.raytracer.objects.mesh.MeshTriangle;

/**
 * Created by IntelliJ IDEA.
 * User: Jörn Dinkla
 * Date: 16.05.2010
 * Time: 10:04:22
 * To change this template use File | Settings | File Templates.
 */
public class SmoothMeshTriangle extends MeshTriangle {

    public SmoothMeshTriangle(final Mesh mesh) {
        super(mesh);
    }

    public SmoothMeshTriangle(final Mesh mesh, final int i0, final int i1, final int i2) {
        super(mesh, i0, i1, i2);
    }

    @Override
    public boolean hit(Ray ray, Hit sr) {
        Point3D v0 = mesh.vertices.get(index0);
        Point3D v1 = mesh.vertices.get(index1);
        Point3D v2 = mesh.vertices.get(index2);

        double a = v0.getX() - v1.getX(), b = v0.getX() - v2.getX(), c = ray.getD().getX(), d = v0.getX() - ray.getO().getX();
        double e = v0.getY() - v1.getY(), f = v0.getY() - v2.getY(), g = ray.getD().getY(), h = v0.getY() - ray.getO().getY();
        double i = v0.getZ() - v1.getZ(), j = v0.getZ() - v2.getZ(), k = ray.getD().getZ(), l = v0.getZ() - ray.getO().getZ();

        double m = f * k - g * j, n = h * k - g * l, p = f * l - h * j;
        double q = g * i - e * k, s = e * j - f * i;

        double invDenom = 1.0 / (a * m + b * q + c * s);

        double e1 = d * m - b * n - c * p;
        double beta = e1 * invDenom;

        if (beta < 0.0)
            return (false);

        double r = e * l - h * i;
        double e2 = a * n + d * q + c * r;
        double gamma = e2 * invDenom;

        if (gamma < 0.0)
            return (false);

        if (beta + gamma > 1.0)
            return (false);

        double e3 = a * p - b * r + d * s;
        double t = e3 * invDenom;

        if (t < MathUtils.INSTANCE.getK_EPSILON())
            return (false);

        sr.setT(t);
        sr.setNormal(interpolateNormal(beta, gamma)); // for smooth shading
        //sr.localHitPoint = ray.linear(t);

        return (true);
    }

    protected Normal interpolateNormal(final double beta, final double gamma) {
        Vector3D v1 = mesh.normals.get(index0).mult(1 - beta - gamma);
        Vector3D v2 = mesh.normals.get(index1).mult(beta);
        Vector3D v3 = mesh.normals.get(index2).mult(gamma);
        Normal normal = new Normal(v1.plus(v2).plus(v3));
        return normal.normalize();
    }

}

File: src/net/dinkla/raytracer/objects/mesh/SmoothMeshTriangle.kt
package net.dinkla.raytracer.objects.mesh

import net.dinkla.raytracer.hits.Hit
import net.dinkla.raytracer.math.*
import net.dinkla.raytracer.objects.mesh.Mesh
import net.dinkla.raytracer.objects.mesh.MeshTriangle

/**
 * Created by IntelliJ IDEA.
 * User: Jörn Dinkla
 * Date: 16.05.2010
 * Time: 10:04:22
 * To change this template use File | Settings | File Templates.
 */
class SmoothMeshTriangle : MeshTriangle {

    constructor(mesh: Mesh) : super(mesh) {}

    constructor(mesh: Mesh, i0: Int, i1: Int, i2: Int) : super(mesh, i0, i1, i2) {}

    override fun hit(ray: Ray, sr: Hit): Boolean {
        val v0 = mesh.vertices[index0]
        val v1 = mesh.vertices[index1]
        val v2 = mesh.vertices[index2]

        val a = v0.x - v1.x
        val b = v0.x - v2.x
        val c = ray.d.x
        val d = v0.x - ray.o.x
        val e = v0.y - v1.y
        val f = v0.y - v2.y
        val g = ray.d.y
        val h = v0.y - ray.o.y
        val i = v0.z - v1.z
        val j = v0.z - v2.z
        val k = ray.d.z
        val l = v0.z - ray.o.z

        val m = f * k - g * j
        val n = h * k - g * l
        val p = f * l - h * j
        val q = g * i - e * k
        val s = e * j - f * i

        val invDenom = 1.0 / (a * m + b * q + c * s)

        val e1 = d * m - b * n - c * p
        val beta = e1 * invDenom

        if (beta < 0.0)
            return false

        val r = e * l - h * i
        val e2 = a * n + d * q + c * r
        val gamma = e2 * invDenom

        if (gamma < 0.0)
            return false

        if (beta + gamma > 1.0)
            return false

        val e3 = a * p - b * r + d * s
        val t = e3 * invDenom

        if (t < MathUtils.K_EPSILON)
            return false

        sr.t = t
        sr.normal = interpolateNormal(beta, gamma) // for smooth shading
        //sr.localHitPoint = ray.linear(t);

        return true
    }

    protected fun interpolateNormal(beta: Double, gamma: Double): Normal {
        val v1 = mesh.normals[index0].mult(1.0 - beta - gamma)
        val v2 = mesh.normals[index1].mult(beta)
        val v3 = mesh.normals[index2].mult(gamma)
        val normal = Normal(v1.plus(v2).plus(v3))
        return normal.normalize()
    }

}
